### ISSUE TITLE
Discussion: Use SDL_ttf for font rendering

### ISSUE THREAD

**Issue by tobbi** [SuperTux/supertux#130]
I *think* it might be more efficient, and we can ditch a lot of code in the process if we were to ditch our own font rendering code for SDL_ttf. 

I think this would require a rewrite of our rendering code and converting the font files to ttf.

**Comment by Karkus476** [comment]
May it fix https://github.com/SuperTux/supertux/issues/97 
?

**Comment by tobbi** [comment]
It might. It might fix #94 as well.

**Comment by tobbi** [comment]
Karkus476: If you want to be helpful, you may try to get fontforge to load and create ttf files out of the data/engine/fonts directory... I couldn't get it to run on my Mac.

**Comment by Karkus476** [comment]
Yeah, that makes it sound like a pretty good idea. If it's done we need to *update the website* so people know to get that as a dependancy

**Comment by tobbi** [comment]
Yes, but that will be done afterwards. After you created the ttf fonts, I can try to play around with SDL_ttf in order to get things going. I'm not sure how much of a trouble this is.

**Comment by Karkus476** [comment]
Okay, I'll see if I can create the font files

**Comment by Karkus476** [comment]
[Font forge](http://fontforge.github.io/en-US/) can't open the .png or the .stf files.

**Comment by maxteufel** [comment]
> can't open the .png

Well, TTF and all other common font file types (OTF, WOFF) use [vector graphics](https://en.wikipedia.org/wiki/Vector_graphics) to ensure the font looks good even when using a very big size.

> .stf

These files are used by SuperTux's custom font rendering system (and are afaik not used by any other software).

This means one would have to re-draw the letters as vector graphics (or use other open-source fonts). Furthermore, TTF doesn't support (multiple) colors inside one single letter ([OTF does](https://en.wikipedia.org/wiki/OpenType#Color)).

**Comment by Karkus476** [comment]
Yeah I thought TTF might use vector graphics

**Comment by tobbi** [comment]
You can find the glyphs in PNG format here:
data/images/engine/fonts/

**Comment by Karkus476** [comment]
You can't convert png to ttf!

**Comment by tobbi** [comment]
I was hoping that FontForge would be able to open the glyphs, vector-trace them and then create a ttf out of this.

**Comment by maxteufel** [comment]
You could use Inkscape to [trace the PNGs](https://inkscape.org/en/doc/tracing/tutorial-tracing.html) and then try to import it in FontForge.

**Comment by Karkus476** [comment]
tobbi: Not in fontforge

**Comment by Karkus476** [comment]
Well I've traced the png but font forge is stubborn and wont let me import it. It's not as simple as just tracing the bitmap, all the letters need to be in the correct format.

**Comment by maxteufel** [comment]
Already looked at <http://designwithfontforge.com/en-US/Importing_Glyphs_from_Other_Programs.html>

**Comment by Karkus476** [comment]
I can't do it, someone else can try, or we could use a placement ttf font for testing purposes until we can create/convert a proper one.

**Comment by Karkus476** [comment]
@maxteufel You seem to know what you're doing, could you try to do it?

**Comment by giby** [comment]
I'll do that at some point.... please keep the original font... ST respects the work of earlier dev...

**Comment by Mathnerd314** [comment]
@Grumbel Is the console font (andale10.png, andale12.png, console2.png) [Andale Mono](https://en.wikipedia.org/wiki/Andale_Mono)? We're probably not allowed to distribute it if that's the case...

**Comment by Grumbel** [comment]
@Mathnerd314 Yes, that's Andale Mono, font copyright however allows that kind of stuff (kind of, maybe). We can replace that with Inconsolata or any of the other free monospace fonts around.

As for the SuperTux font, I would keep that bitmap, it looks nice and works as is. With TTF colors, outlines, shadows and stuff get a lot more complicated and you no longer can do font animations.

**Comment by Flakebi** [comment]
supertuxkart/stk-code#2311 is maybe interesting here too. It would be interesting to know the performance impact, although it would differ for st, it can give some hints.

**Comment by maxteufel** [comment]
Some work has been done in the `feature/modern_fonts_v2` branch using a custom version of SDL_ttf (see the SuperTux/SDL_ttf repository) that uses libraqm (also see comments on #97). The branch will probably need some CMake adjustments to deal with shipping the custom library.

**Comment by Savsish** [comment]
How about fixing Windows MSVC builds

**Comment by tobbi** [comment]
@Savsish Is your comment related to *this* branch only or to other branches. If the latter, open a new issue about it.
